core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java

```diff
@@ -294,21 +294,21 @@ default boolean shouldUseDoubleMultiplication(RelDataTypeFactory typeFactory,
         int s1 = type1.getScale();
         int s2 = type2.getScale();
 
-        final int maxNumericPrecision = getMaxNumericPrecision();
-        int dout =
-            Math.min(
-                p1 - s1 + s2,
-                maxNumericPrecision);
-
+        int d = p1 - s1 + s2;
         int scale = Math.max(6, s1 + p2 + 1);
-        scale =
-            Math.min(
-                scale,
-                maxNumericPrecision - dout);
-        scale = Math.min(scale, getMaxNumericScale());
+        int precision = d + scale;
+
+  // Rules from
+  // https://learn.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql
+        int bound = getMaxNumericPrecision() - 6;  // 32 in the MS documentation
+        if (precision <= bound) {
+          scale = Math.min(scale, getMaxNumericScale() - (precision - scale));
+        } else {
+          // precision > bound
+          scale = Math.min(6, scale);
```

Comment:
If MaxNumericScale < 6, is it possible here to exceed

Reply:
I will fix this case too, but, frankly, if a type system for DECIMAL does not have 6 digits of scale it's probably not very useful

Reply:
+1